Countio

LearnWeaver opened the pull request:
This is a work in progress, see issue #2815 

Any comments or changes needed, please let me know.

tannewt commented on ports/cxd56/mpconfigport.mk:
```diff
@@ -12,6 +12,7 @@ CIRCUITPY_AUDIOIO = 0
 CIRCUITPY_AUDIOBUSIO = 0
 CIRCUITPY_I2CSLAVE = 0
 CIRCUITPY_ROTARYIO = 0
+CIRCUITPY_LINEARIO = 0
```
CIRCUITPY_COUNTIO here and a few other spots

tannewt commented on ports/atmel-samd/common-hal/countio/LinearEncoder.c:
```diff
@@ -0,0 +1,80 @@
+void linearencoder_interrupt_handler(uint8_t channel) {
+    countio_linearencoder_obj_t* self = get_eic_channel_data(channel);
+    uint8_t current_state = gpio_get_pin_level(self->pin_a);
+    if(current_state != self->last_state){
```
Why do you need to check this? Shouldn't we increment every edge?

tannewt commented on shared-bindings/countio/LinearEncoder.c:
```diff
@@ -0,0 +1,141 @@
+
+//| .. currentmodule:: countio
+//|
+//| :class:`LinearEncoder` -- Track the relative position of an linear encoder
```
I don't think this should be called LinearEncoder because it's simply counting the number of edges since startup. I'd just call it `Counter`. It might be worth adding a reset() too.

LearnWeaver commented:
Thanks very much for providing the review and comments. I've now made a few changes:
* Changed to class Counter 
* Fixed (I think) the submodules
* Increment on every edge detection
* Changed the build so that other platforms are excluded (for now)
The build is still failing on the pyruler board (DE language) - any advice on how to get it down by 60 bytes?

tannewt commented on py/circuitpy_mpconfig.mk:
```diff
@@ -109,6 +109,12 @@ CIRCUITPY_DIGITALIO = 1
 CFLAGS += -DCIRCUITPY_DIGITALIO=$(CIRCUITPY_DIGITALIO)
 
+ifndef CIRCUITPY_COUNTIO
+CIRCUITPY_COUNTIO = 1
```
```suggestion
CIRCUITPY_COUNTIO = $(CIRCUITPY_FULL_BUILD)
```

tannewt replied:
This will fix pyruler because it isn't a FULL_BUILD.

LearnWeaver replied:
Thanks that makes sense. I've now changed this.

tannewt commented on py/circuitpy_mpconfig.mk:
```diff
@@ -207,6 +213,11 @@ CIRCUITPY_ROTARYIO = 1
 endif
 CFLAGS += -DCIRCUITPY_ROTARYIO=$(CIRCUITPY_ROTARYIO)
 
+ifndef CIRCUITPY_COUNTIO
+CIRCUITPY_COUNTIO = 1
+endif
+CFLAGS += -DCIRCUITPY_COUNTIO=$(CIRCUITPY_COUNTIO)
+
```
```suggestion
```

dhalbert replied:
As of #2799, simple `ifndefs` now use `?=`, e.g.
https://github.com/adafruit/circuitpython/blob/c311b5c002d004d7a985ecdaf8e0344b35dca209/py/circuitpy_mpconfig.mk#L36-L37

LearnWeaver replied:
No problem, thank you. I've changed this.

tannewt commented on shared-bindings/countio/Counter.c:
```diff
@@ -0,0 +1,152 @@
+//|     
+//|     while True:
+//|         count = pin_counter.count
+//|         if count == None 10:
```
What should this be? The `10` looks like it is extra.

LearnWeaver replied:
It was, sorry!. I've updated the example.

tannewt commented on py/circuitpy_mpconfig.mk:
```diff
@@ -109,6 +109,12 @@ CIRCUITPY_DIGITALIO = 1
 endif
 CFLAGS += -DCIRCUITPY_DIGITALIO=$(CIRCUITPY_DIGITALIO)
 
+ifndef CIRCUITPY_COUNTIO
```
No need for the ifndef with ?=

tannewt commented on shared-bindings/countio/Counter.c:
```diff
@@ -0,0 +1,151 @@
+//| :class:`Counter` -- Track the count of edge transistions (pulses) on a given pin
+//| ====================================================================================
+//|
+//| Counter will keep track of the number of edge transistions (pulses) on a given pin
```
It counts based on falling edges, right? Might want to be specific here.

LearnWeaver commented:
Really appreciate the time and feedback. Thanks

tannewt commented on py/circuitpy_mpconfig.mk:
```diff
@@ -109,6 +109,12 @@ CIRCUITPY_DIGITALIO = 1
 endif
 CFLAGS += -DCIRCUITPY_DIGITALIO=$(CIRCUITPY_DIGITALIO)
 
+ifndef CIRCUITPY_COUNTIO
+CIRCUITPY_COUNTIO = 1
+endif
```
I expected you to delete the `ifndef` and `endif` and leave the `?=` line.
```suggestion
CIRCUITPY_COUNTIO ?= 1
```

tannewt commented on py/circuitpy_mpconfig.mk:
```diff
@@ -207,6 +213,11 @@ CIRCUITPY_ROTARYIO = 1
 endif
 CFLAGS += -DCIRCUITPY_ROTARYIO=$(CIRCUITPY_ROTARYIO)
 
+ifndef CIRCUITPY_COUNTIO
+CIRCUITPY_COUNTIO = $(CIRCUITPY_FULL_BUILD)
+endif
```
```suggestion
CIRCUITPY_COUNTIO ?= $(CIRCUITPY_FULL_BUILD)
```

LearnWeaver replied:
Apologies, thanks for this

tannewt replied:
Why is COUNTIO in here twice? This second time will never apply. (Sorry I didn't ask sooner.)

LearnWeaver replied:
No Problem. Thanks again.

LearnWeaver commented:
Thanks again for your help @tannewt - I've merged in master, looks like it breaks on the stm32f405 - with a protomatter issue. Did I merge in the wrong submodule?
Also the catwan_usbhub fails on the locale DE .. (it builds on my machine using English) - any suggestions on what I can do here?

tannewt commented:
Try moving into `lib/protomatter` and checking out `9f71088d2c32206c6f0495704ae0c040426d5764` which is the commit currently in master. That should hopefully fix it.

For the catwan you can just disable COUNTIO for it. It must be tight on space. You can add `TRANSLATION=de_DE` to the `make` command if you want to try building it locally.

Thanks!
